### PR TITLE
Feature run spmd - add standalone driver program for SPMD work

### PR DESCRIPTION
**Pull request by lessw2020**
This PR adds a driver program, run_spmd.py, that can be used to run SPMD directly in a configurable environment with a given sample model.
Usage is simply "python run_spmd.py" from the spmd directory. 
It sets up and runs with a configurable world size and configurable models.
It then runs the model using both DDP and SPMD, and then does a comparison of the grads returned via allclose as well as a two row visual display of a grad subset. 
run_smpd_results

By default it runs a small 2 layer linear model, with a configurable layer size. 
Also in there is the permute model subset from Rich. 

I've been using it for the fusion work and found it really helpful for being able to quickly run and experiment/verify things. Thus wanted to add it directly to the repo so hopefully it is useful for others by being able to quickly spin up a given model and environment for SPMD.

### CODE REVIEW

**Comment by fegin** [comment]
General comment, we should move this file to an example folder or testing folder but not just reside in the plain SPMD folder as this is the entry point of all the functionality APIs.

**Comment by lessw2020** [comment]
> General comment, we should move this file to an example folder or testing folder but not just reside in the plain SPMD folder as this is the entry point of all the functionality APIs.

Ok moving to /examples/spmd/run_spmd.py

File: examples/spmd/run_spmd.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    # main work
+    work_main(rank, world_size)
+
+    # teardown
+    teardown(rank)
+
+
+if __name__ == "__main__":
+    os.environ["MASTER_ADDR"] = "localhost"
+    os.environ["MASTER_PORT"] = "29502"
```
Comment (fegin) [inline examples/spmd/run_spmd.py:184]: Prefer to have a random number > 1024. Otherwise the same program executing twice may get the port being occupied message.

File: examples/spmd/run_spmd.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        super().__init__()
+        self.w = torch.nn.Parameter(torch.rand((5, 10)))
+        self.b = torch.nn.Parameter(torch.rand((5)))
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x_t = x.permute(0, 2, 1)
+        return torch.nn.functional.linear(x_t, self.w, self.b)
+
+
+class replicaModel(nn.Module):
```
Comment (fegin) [inline examples/spmd/run_spmd.py:91]: nit: `ReplicaModel`. Python naming rule.

**Review by fegin — APPROVED** [review]
LGTM, have some nits.

File: examples/spmd/run_spmd.py (hunk 2 above)
Comment (lessw2020) [reply examples/spmd/run_spmd.py:91]: my work in Rust continues to show lol.
fixed and thanks.

File: examples/spmd/run_spmd.py (hunk 1 above)
Comment (lessw2020) [reply examples/spmd/run_spmd.py]: good idea.  added random port selection with seed. 

Went with the recommended dynamic range for port selection to be safe:

"Dynamic ports—Ports in the range 49152 to 65535 are not assigned, controlled, or registered. They are used for temporary or private ports. They are also known as private or non-reserved ports. Clients should choose ephemeral port numbers from this range..."
